[ICE]: `missing tokens for node`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' ($TID) panicked at compiler/rustc_ast/src/tokenstream.rs:638:54: 'missing tokens for node: Expr { id: NodeId(4294967040), kind: Path(None, Path { span: /tmp/IMM/a.rs:8:12: 8:13 (#0), segments: [PathSegment { ident: x#0, id: NodeId(4294967040), args: None }], tokens: None }), span: /tmp/IMM/a.rs:8:12: 8:13 (#0), attrs: [], tokens: None }'', 'thread 'rustc' ($TID) panicked at compiler/rustc_ast/src/tokenstream.rs:638:54: 'missing tokens for node: Expr { id: NodeId(4294967040), kind: Path(None, Path { span: /tmp/IMM/a.rs:8:12: 8:13 (#0), segments: [PathSegment { ident: x#0, id: NodeId(4294967040), args: None }], tokens: None }), span: /tmp/IMM/a.rs:8:12: 8:13 (#0), attrs: [], tokens: None }''
File: /tmp/IMM/a.rs
-->

auto-reduced (treereduce-rust):
````rust
fn main() {
    macro_rules! m {
        ($g : guard) => {
            m!($g)
        };
    }
    m !(if x)
}

````

original:
````rust
#![feature(macro_guard_matcher)]
fn main() {
    macro_rules! m {
        ($g : guard) => {
            m!($g)
        };
    }
    m !(if x)
}

````

Version information
````
rustc 1.97.0-nightly (bd1e7c794 2026-04-15)
binary: rustc
commit-hash: bd1e7c79485d6a4113bb11dd98cb8b415cd4a55e
commit-date: 2026-04-15
host: x86_64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.2

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/bd1e7c79485d6a4113bb11dd98cb8b415cd4a55e/compiler/rustc_ast/src/tokenstream.rs#L632-L644

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: `guard` fragments in macro are unstable
 --> /tmp/icemaker_global_tempdir.kseR1I47vv8o/rustc_testrunner_tmpdir_reporting.JrF9cRipeMoL/mvce.rs:3:10
  |
3 |         ($g : guard) => {
  |          ^^^^^^^^^^
  |
  = note: see issue #153104 <https://github.com/rust-lang/rust/issues/153104> for more information
  = help: add `#![feature(macro_guard_matcher)]` to the crate attributes to enable
  = note: this compiler was built on 2026-04-15; consider upgrading it if it is out of date


thread 'rustc' (1298682) panicked at /rustc-dev/bd1e7c79485d6a4113bb11dd98cb8b415cd4a55e/compiler/rustc_ast/src/tokenstream.rs:638:54:
missing tokens for node: Expr { id: NodeId(4294967040), kind: Path(None, Path { span: /tmp/icemaker_global_tempdir.kseR1I47vv8o/rustc_testrunner_tmpdir_reporting.JrF9cRipeMoL/mvce.rs:7:12: 7:13 (#0), segments: [PathSegment { ident: x#0, id: NodeId(4294967040), args: None }], tokens: None }), span: /tmp/icemaker_global_tempdir.kseR1I47vv8o/rustc_testrunner_tmpdir_reporting.JrF9cRipeMoL/mvce.rs:7:12: 7:13 (#0), attrs: [], tokens: None }
stack backtrace:
   0:     0x7f41b0024b7b - <<std[c5ab73980a16b942]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[a93ae171b9b17a3a]::fmt::Display>::fmt
   1:     0x7f41b0621448 - core[a93ae171b9b17a3a]::fmt::write
   2:     0x7f41b003be46 - <std[c5ab73980a16b942]::sys::stdio::unix::Stderr as std[c5ab73980a16b942]::io::Write>::write_fmt
   3:     0x7f41afffae78 - std[c5ab73980a16b942]::panicking::default_hook::{closure#0}
   4:     0x7f41b00181e3 - std[c5ab73980a16b942]::panicking::default_hook
   5:     0x7f41aeff0a0c - std[c5ab73980a16b942]::panicking::update_hook::<alloc[73519e9f23c77201]::boxed::Box<rustc_driver_impl[733b6e0b8e0e976b]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f41b00184c2 - std[c5ab73980a16b942]::panicking::panic_with_hook
   7:     0x7f41afffaf38 - std[c5ab73980a16b942]::panicking::panic_handler::{closure#0}
   8:     0x7f41afff1c79 - std[c5ab73980a16b942]::sys::backtrace::__rust_end_short_backtrace::<std[c5ab73980a16b942]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f41afffc99d - __rustc[2880848b2880f319]::rust_begin_unwind
  10:     0x7f41acb4d3bc - core[a93ae171b9b17a3a]::panicking::panic_fmt
  11:     0x7f41b1bdbbc1 - <rustc_expand[144a5cb01134d204]::expand::MacroExpander>::fully_expand_fragment
  12:     0x7f41b1b9631d - <rustc_expand[144a5cb01134d204]::expand::MacroExpander>::expand_crate
  13:     0x7f41b0de818d - rustc_interface[9d19a1b9741214bc]::passes::resolver_for_lowering_raw
  14:     0x7f41b0de75d3 - rustc_query_impl[3291eb9cebafb6fd]::query_impl::resolver_for_lowering_raw::invoke_provider_fn::__rust_begin_short_backtrace
  15:     0x7f41b1771f26 - rustc_query_impl[3291eb9cebafb6fd]::execution::try_execute_query::<rustc_middle[9abe997593166c8f]::query::caches::SingleCache<rustc_middle[9abe997593166c8f]::query::erase::ErasedData<[u8; 16usize]>>, false>
  16:     0x7f41b1754444 - rustc_query_impl[3291eb9cebafb6fd]::query_impl::resolver_for_lowering_raw::execute_query_non_incr::__rust_end_short_backtrace
  17:     0x7f41b175226c - rustc_interface[9d19a1b9741214bc]::interface::run_compiler::<(), rustc_driver_impl[733b6e0b8e0e976b]::run_compiler::{closure#0}>::{closure#1}
  18:     0x7f41b17395be - std[c5ab73980a16b942]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[9d19a1b9741214bc]::util::run_in_thread_with_globals<rustc_interface[9d19a1b9741214bc]::util::run_in_thread_pool_with_globals<rustc_interface[9d19a1b9741214bc]::interface::run_compiler<(), rustc_driver_impl[733b6e0b8e0e976b]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  19:     0x7f41b1739e60 - <std[c5ab73980a16b942]::thread::lifecycle::spawn_unchecked<rustc_interface[9d19a1b9741214bc]::util::run_in_thread_with_globals<rustc_interface[9d19a1b9741214bc]::util::run_in_thread_pool_with_globals<rustc_interface[9d19a1b9741214bc]::interface::run_compiler<(), rustc_driver_impl[733b6e0b8e0e976b]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[a93ae171b9b17a3a]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  20:     0x7f41b173acec - <std[c5ab73980a16b942]::sys::thread::unix::Thread>::new::thread_start
  21:     0x7f41ab26e97a - <unknown>
  22:     0x7f41ab2f22bc - <unknown>
  23:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.97.0-nightly (bd1e7c794 2026-04-15) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [resolver_for_lowering_raw] getting the resolver for lowering
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0658`.

```

</p>
</details>

<!--
query stack:
#0 [resolver_for_lowering_raw] getting the resolver for lowering

-->
@rustbot label +F-macro_guard_matcher

--- a/compiler/rustc_parse/src/parser/expr.rs
+++ b/compiler/rustc_parse/src/parser/expr.rs
@@ -3464,7 +3464,9 @@ impl<'a> Parser<'a> {
     }
 
     pub(crate) fn eat_metavar_guard(&mut self) -> Option<Box<Guard>> {
-        self.eat_metavar_seq(MetaVarKind::Guard, |this| this.parse_match_arm_guard()).flatten()
+        self.eat_metavar_seq(MetaVarKind::Guard, |this| {
+            this.expect_match_arm_guard(ForceCollect::Yes)
+        })
     }
 
     fn parse_match_arm_guard(&mut self) -> PResult<'a, Option<Box<Guard>>> {

--- a/tests/ui/macros/macro-guard-matcher-recursion.rs
+++ b/tests/ui/macros/macro-guard-matcher-recursion.rs
@@ -1,0 +1,10 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/155333>
+#![feature(macro_guard_matcher)]
+fn main() {
+    macro_rules! m {
+        ($g : guard) => {
+            m!($g) //~ ERROR recursion limit reached while expanding `m!`
+        };
+    }
+    m!(if x)
+}

--- a/tests/ui/macros/macro-guard-matcher-recursion.stderr
+++ b/tests/ui/macros/macro-guard-matcher-recursion.stderr
@@ -1,0 +1,14 @@
+error: recursion limit reached while expanding `m!`
+  --> $DIR/macro-guard-matcher-recursion.rs:6:13
+   |
+LL |             m!($g)
+   |             ^^^^^^
+...
+LL |     m!(if x)
+   |     -------- in this macro invocation
+   |
+   = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`macro_guard_matcher_recursion`)
+   = note: this error originates in the macro `m` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: aborting due to 1 previous error
+

